File: 后端篇/disc/第7章/7.2/server/node/controllers/accounts.js
// JS：controllers\accounts.js
"use strict"
import model from "../models/index.js"

// 无论是返回JSON文本，还是HTML文本，
// 都是一样的，只是渲染不同
export default {
  // 登录页面
  "GET /login": async (ctx, next) => {
    const title = "登录"
    ctx.render("accounts/login", { title })
  },

  // 获取已登录用户对象
  "GET /logon": async (ctx, next) => {
    ctx.status = 200
    ctx.body = ctx.session.user
  },

  // 接收登录的post请求
  "POST /login": async (ctx, next) => {
    const { name, passwd } = ctx.request.body
      , db = await model.connect()
    // 如果密码有加密，在查询前将passwd加密，再查询
    const res = await model.execute(db, "select * from account where name=@name && passwd=@passwd limit 1", { name, passwd })
    if (res.length > 0) {
      // role: assistant、administrator
      const u = res[0]
      ctx.session.user = { id: u.id, name: u.name, role: u.role }
      ctx.redirect("/", 302) // 登录后重定向到首页
    } else {
      ctx.render("login", { title: "登录", errMsg: "登录失败，请重试" })
    }
  },

  // 登出
  "GET /logout": async (ctx, next) => {
    ctx.session.user = null
    ctx.redirect("/accounts/login")
  },
}

File: 后端篇/disc/第7章/7.2/server/go/controllers/accounts.go
// Go：controllers\accounts.go
package controllers

import (
	"encoding/json"
	"github.com/kataras/iris/v12"
	"github.com/kataras/iris/v12/mvc"
	"local/models"
	"xorm.io/xorm"
)

// 路由基地址：/accounts
type AccountController struct {
	Ctx iris.Context // HTTP请求上下文对象
	Db  *xorm.Engine
}

func (c *AccountController) BeforeActivation(b mvc.BeforeActivation) {
	// 自定义url路径
	b.Handle("GET", "/logon", "CustomGetLogon")
}

// GET：/accounts/logon
func (c *AccountController) CustomGetLogon() string {
	return c.Ctx.GetCookie("user")
}

// GET：/accounts/login
func (c *AccountController) GetLogin() {
	c.Ctx.ViewLayout(iris.NoLayout) // 不使用默认模板
	c.Ctx.View("accounts/login", map[string]interface{}{"title": "登录"})
}

// POST：/accounts/login
func (c *AccountController) PostLogin() {
	var account models.Account
	account.Name = c.Ctx.FormValue("name")
	account.Passwd = c.Ctx.FormValue("passwd")

	if account.Name == "" || account.Passwd == "" {
		c.Ctx.ViewLayout(iris.NoLayout) // 不使用默认布局
		c.Ctx.View("accounts/login", map[string]interface{}{"title": "登录", "errMsg": "参数错误"})
	}
	if has, _ := c.Db.Where("name = ? and passwd = ?", account.Name, account.Passwd).Get(&account); has {
		account.Passwd = ""
		userJson, _ := json.Marshal(account)
		c.Ctx.SetCookieKV("user", string(userJson))
		c.Ctx.Redirect("/")
	} else {
		c.Ctx.ViewLayout(iris.NoLayout) // 不使用默认布局
		c.Ctx.View("accounts/login", map[string]interface{}{"title": "登录", "errMsg": "登录失败，请重试"})
	}
}

// GET：/accounts/logout
func (c *AccountController) GetLogout() {
	c.Ctx.RemoveCookie("user")
	c.Ctx.Redirect("/accounts/login")
}
